trimming: Restore JLL support

topolarity opened the pull request:
Add an extra indirection to our printing code for `@assert` so that that inferrability is restored after the override. Also partially revert d77c24f009b93577bcf254908e639c9f149f2e4e which, deleted init code important for JLL's

~This still requires that we bypass this inference-disabling code in JLLWrappers.jl:~
```julia
if isdefined(Base, :Experimental) && isdefined(Base.Experimental, Symbol("@compiler_options"))
    @eval Base.Experimental.@compiler_options compile=min optimize=0 infer=false
end
```

topolarity committed:
trimming: Restore JLL support

Add an extra indirection to our printing code for `@assert` so that
that inferrability is restored after the override. Also partially
revert d77c24f009b93577bcf254908e639c9f149f2e4e which, deleted init
code important for JLL's

This still requires that we bypass this inference-disabling code in
JLLWrappers.jl:
```julia
if isdefined(Base, :Experimental) && isdefined(Base.Experimental, Symbol("@compiler_options"))
    @eval Base.Experimental.@compiler_options compile=min optimize=0 infer=false
end
```

topolarity committed:
Fix `Base.Experimental.@compiler_options` doc-string

JeffBezanson commented:
Does the PackageCompiler artifact bundling system solve this? I feel like we should be using that.

topolarity committed:
trimming: Add basic JLL test

This gets us JLL coverage in our trimming tests.

This test project should be re-worked to actually use `juliac.jl`
instead of re-implementing it out-of-place, and also to make it easier
to add additional tests in the future.

topolarity commented:
> Does the PackageCompiler artifact bundling system solve this? I feel like we should be using that.

I don't believe so - `Base.DEPOT_PATH` still needs to get initialized so that we can find the depot PackageCompiler would have prepared for us

topolarity commented:
Found a way to re-enable inference for JLL's

We no longer need a branch, and this now adds a test for it 👍

JeffBezanson commented:
OK, I think what we can do is use a much simpler version of `init_depot_path` when using a bundled depot, that just sets it to the path relative to the exe location.

topolarity commented on contrib/juliac-buildscript.jl:
```diff
@@ -271,6 +268,12 @@ let loaded = Symbol.(Base.loaded_modules_array())  # TODO better way to do this
+    if :JLLWrappers in loaded
+        using JLLWrappers
+        @eval JLLWrappers begin
+            Base.Experimental.@compiler_options compile=min optimize=2 infer=true
```
This should no longer be necessary once https://github.com/JuliaLang/julia/pull/57653 lands
